src/evaluators/BinaryExpression.js

```diff
@@ -40,38 +40,40 @@ let unknownValueOfOrToString = "might be an object with an unknown valueOf or to
 export function getPureBinaryOperationResultType(
   realm: Realm, op: BabelBinaryOperator, lval: Value, rval: Value, lloc: ?BabelNodeSourceLocation, rloc: ?BabelNodeSourceLocation
 ): void | typeof Value {
+  function reportErrorIfNotPure(purityTest: (Realm, Value) => boolean, typeIfPure: typeof Value): void | typeof Value {
+    let leftPure = purityTest(realm, lval);
+    let rightPure = purityTest(realm, rval);
+    if (leftPure && rightPure) return BooleanValue;
```

Comment:
We would need a test case that depends on the type of the abstract value being a number rather than a Boolean. Right now that would be quite tricky to set up.